src/util/binary.c

```diff
@@ -0,0 +1,72 @@
+/* 
```

Comment:
Hi there,

I think you can rewrite your singularity_binary_arch() function like this, so that the only requirement is to include the system include file 'elf.h' and avoid adding a new dependency 'libelf-dev'. This would shorten and simplify this patch quite a bit.
 
```
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>

#include <elf.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define BINARY_SYSTEM_ERROR -1

int
singularity_binary_arch(char *path)
{
	int fd, ret = BINARY_ARCH_UNKNOWN;
	char *elf;
	struct stat st;

	/* prepare file */
	if(path == NULL){
		fprintf(stderr, "Error invalid filename\n");
		return -1;
	}
	fd = open(path, O_RDONLY);
	if(fd < 0){
		perror("Error while opening file");
		return -1;
	}
	if(fstat(fd, &st) < 0){
		perror("Error calling fstat()");
		close(fd);
		return -1;
	}
	elf = mmap(NULL, st.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
	if(elf == MAP_FAILED){
		perror("Error mapping file");
		close(fd);
		return -1;
	}

	/* file is ready for inspection */
	if(memcmp(elf, ELFMAG, strlen(ELFMAG)) != 0){
		fprintf(stderr, "Not an ELF file\n");
	}else if(elf[EI_CLASS] == ELFCLASS32){
		Elf32_Ehdr *hdr = elf;
		if(hdr->e_machine == EM_386)
			ret = BINARY_ARCH_I386;
		if(hdr->e_machine == EM_X86_64)
			ret = BINARY_ARCH_X32;
	}else if(elf[EI_CLASS] == ELFCLASS64){
		Elf64_Ehdr *hdr = elf;
		if(hdr->e_machine == EM_X86_64)
			ret = BINARY_ARCH_X86_64;
	}

	/* cleanup */
	munmap(elf, st.st_size);
	close(fd);

	return ret;
}
```